remote.go

```diff
@@ -755,14 +759,28 @@ func doCalculateRefs(
 
 func getWants(localStorer storage.Storer, refs memory.ReferenceStorage) ([]plumbing.Hash, error) {
 	wants := map[plumbing.Hash]bool{}
+
+	shallows, err := localStorer.Shallow()
+	if err != nil {
+		return nil, err
+	}
+	shallowCommits := map[plumbing.Hash]bool{}
+	for _, shallow := range shallows {
+		shallowCommits[shallow] = true
+	}
+
 	for _, ref := range refs {
 		hash := ref.Hash()
 		exists, err := objectExists(localStorer, ref.Hash())
 		if err != nil {
 			return nil, err
 		}
 
-		if !exists {
+		if exists {
+			if isShallow, ok := shallowCommits[hash]; ok && isShallow {
```

Comment:
nit: I think at that point you can directly simplify this by `_, isShallow := shallowCommits[hash]; isShallow`. The value will always be true if the void is found in the map.